bson/src/main/org/bson/internal/LazyCodec.java

```diff
@@ -23,14 +23,19 @@
 import org.bson.codecs.EncoderContext;
 import org.bson.codecs.configuration.CodecRegistry;
 
+import java.lang.reflect.Type;
+import java.util.List;
+
 class LazyCodec<T> implements Codec<T> {
     private final CodecRegistry registry;
     private final Class<T> clazz;
+    private final List<Type> types;
```

Comment:
By making LazyCodec aware of type arguments, it can properly get the parameterized codec when it need to.